tests/unit/test_openstack_manager.py

```diff
@@ -87,6 +113,203 @@ def test_list_projects_missing_credentials_error(
+
+LANG=C.UTF-8
+ACTIONS_RUNNER_HOOK_JOB_STARTED=
+""",
+            id="dockerhub mirror set",
+        ),
+        pytest.param(
+            None,
+            None,
+            [
+                openstack_manager.SSHDebugConnection(
+                    host="127.0.0.1",
+                    port=10022,
+                    rsa_fingerprint="SHA256:testrsa",
+                    ed25519_fingerprint="SHA256:tested25519",
+                )
+            ],
+            """PATH=/home/ubuntu/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
+
+
+
+
+
+LANG=C.UTF-8
+ACTIONS_RUNNER_HOOK_JOB_STARTED=
+
+TMATE_SERVER_HOST=127.0.0.1
+TMATE_SERVER_PORT=10022
+TMATE_SERVER_RSA_FINGERPRINT=SHA256:testrsa
+TMATE_SERVER_ED25519_FINGERPRINT=SHA256:tested25519
+""",
+            id="ssh debug connection set",
+        ),
+        pytest.param(
+            openstack_manager.ProxyConfig(
+                http="http://test.internal",
+                https="https://test.internal",
+                no_proxy="http://no_proxy.internal",
+            ),
+            "http://dockerhub_mirror.test",
+            [
+                openstack_manager.SSHDebugConnection(
+                    host="127.0.0.1",
+                    port=10022,
+                    rsa_fingerprint="SHA256:testrsa",
+                    ed25519_fingerprint="SHA256:tested25519",
+                )
+            ],
+            """PATH=/home/ubuntu/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
+
+HTTP_PROXY=http://test.internal
+http_proxy=http://test.internal
+
+
+HTTPS_PROXY=https://test.internal
+https_proxy=https://test.internal
+
+
+
+NO_PROXY=http://no_proxy.internal
+no_proxy=http://no_proxy.internal
+
+
+DOCKERHUB_MIRROR=http://dockerhub_mirror.test
+CONTAINER_REGISTRY_URL=http://dockerhub_mirror.test
+
+LANG=C.UTF-8
+ACTIONS_RUNNER_HOOK_JOB_STARTED=
+
+TMATE_SERVER_HOST=127.0.0.1
+TMATE_SERVER_PORT=10022
+TMATE_SERVER_RSA_FINGERPRINT=SHA256:testrsa
+TMATE_SERVER_ED25519_FINGERPRINT=SHA256:tested25519
+""",
+            id="all values set",
+        ),
+    ],
+)
+def test__generate_runner_env(
+    proxy_config: Optional[openstack_manager.ProxyConfig],
+    dockerhub_mirror: Optional[str],
+    ssh_debug_connections: Optional[list[openstack_manager.SSHDebugConnection]],
+    expected_env_contents: str,
+):
+    """
+    arrange: given configuration values for runner environment.
+    act: when _generate_runner_env is called.
+    assert: expected .env contents are generated.
+    """
+    environment = jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True)
+    assert (
+        openstack_manager._generate_runner_env(
+            templates_env=environment,
+            proxies=proxy_config,
+            dockerhub_mirror=dockerhub_mirror,
+            ssh_debug_connections=ssh_debug_connections,
+        )
+        == expected_env_contents
+    )
+
+
 def test__build_image_command():
```

Comment:
```suggestion
def test_build_image_command():
```

Reply:
This is a test for an internal function `_build_image_command`, and there is a fct `build_image` so i think this is alright :)